### ISSUE TITLE
Log to remote FS azure fails when LightningModule uses the `self.log` method

### ISSUE THREAD

**Issue by ioangatop** [Lightning-AI/lightning#18445]
### Bug description

Hi all! While trying to log to a remote file system, specifically in Azure, I found that when my `LightningModule` uses the `self.log` method it fails at the end of the first epoch. Removing the `TensorBoard` logger, this is where the error seemed to be generated from, also did not help as it fails to save the checkpoint of the model.

Strangely, when the `LightningModule` does not use the `self.log`, it works perfectly (create files, saves config, tfevent and chackpoints).

The following, for example, works as expected (for the version that is not, please see bellow):
```yaml
# main.py
# ------------------------------------
# from pytorch_lightning import cli
# def main() -> None:
#     cli.LightningCLI()
# 
# if __name__ == "__main__":
#     main()
# ------------------------------------
# >>> export FSSPEC_ABFS='{"anon": false}'
# >>> python main.py fit --config config.yaml
---
seed_everything: 42
trainer:
  default_root_dir: &OUTPUT_DIR az://<container-name>@<name>.blob.core.windows.net/tmp/
  max_epochs: 10
  logger:
    class_path: TensorBoardLogger
    init_args:
      save_dir: *OUTPUT_DIR
model:
  class_path: pytorch_lightning.demos.boring_classes.BoringModel
data:
  class_path: pytorch_lightning.demos.boring_classes.BoringDataModule
```

### What version are you seeing the problem on?

master

### How to reproduce the bug

```python
Configuration File:

---
seed_everything: 42
trainer:
  default_root_dir: &OUTPUT_DIR az://<container-name>@<name>.blob.core.windows.net/tmp/
  max_epochs: 10
  logger:
    class_path: TensorBoardLogger
    init_args:
      save_dir: *OUTPUT_DIR
model:
  class_path: model.TestModel
data:
  class_path: pytorch_lightning.demos.boring_classes.BoringDataModule


Model definition:
```python3
# model.py
from pytorch_lightning.demos import boring_classes

class TestModel(boring_classes.BoringModel):
    def training_step(self, batch, batch_idx):
        loss = self.step(batch)
        self.log("train/StepLoss", loss, prog_bar=True)
        return {"loss": loss}
```

Entry point:
```python3
# main.py
"""
>>> export FSSPEC_ABFS='{"anon": false}'
>>> python main.py fit --config config.yaml
"""
from pytorch_lightning import cli

def main() -> None:
    cli.LightningCLI()

if __name__ == "__main__":
    main()
```
```


### Error messages and logs

```
# The blob type is invalid for this operation.
```


### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
        - GPU:               None
        - available:         False
        - version:           None
* Lightning:
        - lightning:         2.1.0rc0
        - lightning-cloud:   0.5.37
        - lightning-utilities: 0.9.0
        - pytorch-lightning: 2.0.7
        - torch:             2.0.1
        - torchmetrics:      1.1.1
* Packages:
        - absl-py:           1.4.0
        - adlfs:             2023.8.0
        - aiohttp:           3.8.5
        - aiosignal:         1.3.1
        - annotated-types:   0.5.0
        - anyio:             4.0.0
        - arrow:             1.2.3
        - async-timeout:     4.0.3
        - attrs:             23.1.0
        - azure-core:        1.29.3
        - azure-datalake-store: 0.0.53
        - azure-identity:    1.12.0
        - azure-storage-blob: 12.17.0
        - backoff:           2.2.1
        - beautifulsoup4:    4.12.2
        - blessed:           1.20.0
        - bs4:               0.0.1
        - cachetools:        5.3.1
        - certifi:           2023.7.22
        - cffi:              1.15.1
        - charset-normalizer: 3.2.0
        - click:             8.1.6
        - croniter:          1.4.1
        - cryptography:      41.0.3
        - dateutils:         0.6.12
        - deepdiff:          6.3.1
        - docstring-parser:  0.15
        - fastapi:           0.99.1
        - filelock:          3.12.3
        - frozenlist:        1.4.0
        - fsspec:            2023.6.0
        - google-auth:       2.22.0
        - google-auth-oauthlib: 1.0.0
        - grpcio:            1.57.0
        - h11:               0.14.0
        - idna:              3.4
        - importlib-resources: 6.0.1
        - iniconfig:         2.0.0
        - inquirer:          3.1.3
        - isodate:           0.6.1
        - itsdangerous:      2.1.2
        - jinja2:            3.1.2
        - jsonargparse:      4.24.0
        - lightning:         2.1.0rc0
        - lightning-cloud:   0.5.37
        - lightning-utilities: 0.9.0
        - markdown:          3.4.4
        - markdown-it-py:    3.0.0
        - markupsafe:        2.1.3
        - mdurl:             0.1.2
        - mpmath:            1.3.0
        - msal:              1.23.0
        - msal-extensions:   1.0.0
        - multidict:         6.0.4
        - networkx:          3.1
        - numpy:             1.25.2
        - oauthlib:          3.2.2
        - ordered-set:       4.1.0
        - packaging:         23.1
        - pip:               23.2.1
        - pluggy:            1.3.0
        - portalocker:       2.7.0
        - protobuf:          4.24.2
        - psutil:            5.9.5
        - pyasn1:            0.5.0
        - pyasn1-modules:    0.3.0
        - pycparser:         2.21
        - pydantic:          1.10.12
        - pydantic-core:     2.6.3
        - pygments:          2.16.1
        - pyjwt:             2.8.0
        - pytest:            7.4.0
        - python-dateutil:   2.8.2
        - python-editor:     1.0.4
        - python-multipart:  0.0.6
        - pytorch-lightning: 2.0.7
        - pytz:              2023.3
        - pyyaml:            6.0.1
        - readchar:          4.0.5
        - requests:          2.31.0
        - requests-oauthlib: 1.3.1
        - rich:              13.5.2
        - rsa:               4.9
        - setuptools:        65.5.0
        - six:               1.16.0
        - sniffio:           1.3.0
        - soupsieve:         2.4.1
        - starlette:         0.27.0
        - starsessions:      1.3.0
        - sympy:             1.12
        - tensorboard:       2.14.0
        - tensorboard-data-server: 0.7.1
        - torch:             2.0.1
        - torchmetrics:      1.1.1
        - tqdm:              4.66.1
        - traitlets:         5.9.0
        - typeshed-client:   2.3.0
        - typing-extensions: 4.7.1
        - urllib3:           1.26.16
        - uvicorn:           0.23.2
        - wcwidth:           0.2.6
        - websocket-client:  1.6.1
        - websockets:        11.0.3
        - werkzeug:          2.3.7
        - wheel:             0.41.2
        - yarl:              1.9.2
* System:
        - OS:                Darwin
        - architecture:
                - 64bit
                - 
        - processor:         arm
        - python:            3.11.5
        - release:           22.6.0
        - version:           Darwin Kernel Version 22.6.0: Wed Jul  5 22:22:05 PDT 2023; root:xnu-8796.141.3~6/RELEASE_ARM64_T6000

</details>

### More info

_No response_

**Comment by ioangatop** [comment]
Found a workaround:
```py3
# pip install pytorch_lightning adlfs tensorboard
# main.py
import pytorch_lightning as pl
from pytorch_lightning.demos import boring_classes

OUTPUT_DIR = "az://<container-name>@<name>.blob.core.windows.net/tmp/"

class TestModel(boring_classes.BoringModel):
    def training_step(self, batch, batch_idx):
        loss = self.step(batch)
        self.log("train/StepLoss", loss, prog_bar=True)  # <-- this is the line that causes the issue
        return {"loss": loss}

model = TestModel()
trainer = pl.Trainer(
    max_epochs=10,
    default_root_dir=OUTPUT_DIR,
    logger=[],
)
trainer.fit(model)
```

**Comment by nkaenzig** [comment]
I'm having the same issue (`lightning==2.2.1`). The above workaround also doesn't work anymore:

```
RequestId:6e66d9c5-d01e-0033-375a-8670de000000
Time:2024-04-04T06:33:45.2051891Z
ErrorCode:InvalidBlobType
Content: <?xml version="1.0" encoding="utf-8"?><Error><Code>InvalidBlobType</Code><Message>The blob type is invalid for this operation.
RequestId:6e66d9c5-d01e-0033-375a-8670de000000
Time:2024-04-04T06:33:45.2051891Z</Message></Error>
```
